src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java

```diff
@@ -0,0 +1,416 @@
+    }
+
+    @SuppressWarnings("PMD.NullAssignment")
+    private void connectionCleanup() {
+        // Must synchronize since we're messing with the shared connection object and this block
+        // is executed in some other thread
+        synchronized (this) {
+            if (client != null) {
+                client = null;
+            }
+        }
+    }
+
+    /**
+     * Run re-subscription task in another thread so that the current thread is not blocked by it. The task will keep
+     * retrying until all subscription succeeded or it's canceled by network interruption.
+     *
+     * @param sessionPresent whether the session persisted
+     */
+    private synchronized void resubscribe(boolean sessionPresent) {
+        // No need to resub if we haven't subscribed to anything
+        if (!subscriptionTopics.isEmpty()) {
+            // If connected without a session, all subscriptions are dropped and need to be resubscribed
+            if (!sessionPresent) {
+                droppedSubscriptionTopics.addAll(subscriptionTopics);
+            }
+            if (!droppedSubscriptionTopics.isEmpty() && (resubscribeFuture == null || resubscribeFuture.isDone())) {
+                resubscribeFuture = executorService.submit(this::resubscribeDroppedTopicsTask);
+            }
+        }
+    }
+
+    private void resubscribeDroppedTopicsTask() {
+        long delayMillis = 0;  // don't delay the first run
+        while (connected() && !droppedSubscriptionTopics.isEmpty()) {
+            logger.atDebug().event(RESUB_LOG_EVENT).kv("droppedTopics",
+                            (Supplier<List<String>>) () -> droppedSubscriptionTopics.stream().map(Subscribe::getTopic)
+                                    .collect(Collectors.toList())).kv("delayMillis", delayMillis)
+                    .log("Subscribing to dropped topics");
+            ScheduledFuture<?> scheduledFuture = ses.schedule(() -> {
+                List<CompletableFuture<SubscribeResponse>> subFutures = new ArrayList<>();
+                for (Subscribe sub : droppedSubscriptionTopics) {
+                    subFutures.add(subscribe(sub).whenComplete((result, error) -> {
+                        if (error == null) {
+                            droppedSubscriptionTopics.remove(sub);
+                        } else {
+                            logger.atError().event(RESUB_LOG_EVENT).cause(error).kv(TOPIC_KEY, sub.getTopic())
+                                    .log("Failed to subscribe to topic. Will retry later");
+                        }
+                    }));
+                }
+                // Block and wait for all subscriptions to finish
+                CompletableFuture<?> allSubFutures =
+                        CompletableFuture.allOf(subFutures.toArray(new CompletableFuture[0]));
+                try {
+                    allSubFutures.get();
+                } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)